File: artifacts/tulipindicators/translate/src/fosc.rs
use crate::*;

pub fn ti_fosc_start(options: Option<&[f64]>) -> i32 {
    // Check if options is None (equivalent to NULL pointer check in C)
    // and if the slice has at least one element
    match options {
        Some(opts) if !opts.is_empty() => opts[0] as i32,
        _ => 0, // Default return value when options is None or empty
    }
}
pub fn helper_ti_fosc_1(
    output_idx_ref: &mut u32,
    size: usize,
    input: &[f64],
    period: i32,
    output: &mut [f64],
) {
    let mut output_idx = *output_idx_ref;
    let mut x = 0.0;
    let mut x2 = 0.0;
    let mut y = 0.0;
    let mut xy = 0.0;
    let p = 1.0 / period as f64;
    let mut tsf = 0.0;

    for i in 0..(period - 1) as usize {
        let i_plus_1 = (i + 1) as f64;
        x += i_plus_1;
        x2 += i_plus_1 * i_plus_1;
        xy += input[i] * i_plus_1;
        y += input[i];
    }

    let period_f64 = period as f64;
    x += period_f64;
    x2 += period_f64 * period_f64;
    let bd = 1.0 / ((period_f64 * x2) - (x * x));

    for i in (period - 1) as usize..size {
        xy += input[i] * period_f64;
        y += input[i];
        let b = ((period_f64 * xy) - (x * y)) * bd;

        {
            let a = (y - (b * x)) * p;
            if i >= period as usize {
                output[output_idx as usize] = (100.0 * (input[i] - tsf)) / input[i];
                output_idx += 1;
            }
            tsf = a + (b * (period_f64 + 1.0));
        }

        xy -= y;
        y -= input[(i - (period as usize)) + 1];
    }

    *output_idx_ref = output_idx;
}
pub fn ti_fosc(
    size: usize,
    inputs: &[&[f64]],
    options: &[f64],
    outputs: &mut [&mut [f64]],
) -> Result<(), i32> {
    let input = inputs[0];
    let period = options[0] as i32;
    let output = &mut outputs[0];
    let mut output_idx = 0;

    if period < 1 {
        return Err(1);
    }

    let start = ti_fosc_start(Some(options));
    if size <= start as usize {
        return Ok(());
    }

    helper_ti_fosc_1(&mut output_idx, size, input, period, output);

    debug_assert!(
        (output[output_idx as usize] - outputs[0][0]) == (size as f64 - start as f64)
    );

    Ok(())
}

File: artifacts/tulipindicators/decompose/src/fosc.c
int ti_fosc_start(const double *options)
{
  return (int) options[0];
}


/*** DEPENDENCIES:
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int']
void helper_ti_fosc_1(unsigned int * const output_idx_ref, int size, const double * const input, const int period, double * const output)
{
  unsigned int output_idx = *output_idx_ref;
  double x = 0;
  double x2 = 0;
  double y = 0;
  double xy = 0;
  const double p = 1.0 / period;
  double tsf = 0;
  ;
  int i;
  for (i = 0; i < (period - 1); i += 1)
  {
    x += i + 1;
    x2 += (i + 1) * (i + 1);
    xy += input[i] * (i + 1);
    y += input[i];
  }

  x += period;
  x2 += period * period;
  const double bd = 1.0 / ((period * x2) - (x * x));
  for (i = period - 1; i < size; i += 1)
  {
    xy += input[i] * period;
    y += input[i];
    const double b = ((period * xy) - (x * y)) * bd;
    do
    {
      const double a = (y - (b * x)) * p;
      if (i >= period)
      {
        output[output_idx] = (100 * (input[i] - tsf)) / input[i];
        output_idx += 1;
      }
      tsf = a + (b * (period + 1));
    }
    while (0);
    xy -= y;
    y -= input[(i - period) + 1];
  }

  *output_idx_ref = output_idx;
}


/*** DEPENDENCIES:
***/


int ti_fosc(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_fosc_start(options))
  {
    return 0;
  }
  do
  {
    helper_ti_fosc_1(&output_idx, size, input, period, output);
  }
  while (0);
  assert(((&output[output_idx]) - outputs[0]) == (size - ti_fosc_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_fosc_start(const double *options)
{
  return (int) options[0];
}


----------------------------
void helper_ti_fosc_1(unsigned int * const output_idx_ref, int size, const double * const input, const int period, double * const output)
{
  unsigned int output_idx = *output_idx_ref;
  double x = 0;
  double x2 = 0;
  double y = 0;
  double xy = 0;
  const double p = 1.0 / period;
  double tsf = 0;
  ;
  int i;
  for (i = 0; i < (period - 1); i += 1)
  {
    x += i + 1;
    x2 += (i + 1) * (i + 1);
    xy += input[i] * (i + 1);
    y += input[i];
  }

  x += period;
  x2 += period * period;
  const double bd = 1.0 / ((period * x2) - (x * x));
  for (i = period - 1; i < size; i += 1)
  {
    xy += input[i] * period;
    y += input[i];
    const double b = ((period * xy) - (x * y)) * bd;
    do
    {
      const double a = (y - (b * x)) * p;
      if (i >= period)
      {
        output[output_idx] = (100 * (input[i] - tsf)) / input[i];
        output_idx += 1;
      }
      tsf = a + (b * (period + 1));
    }
    while (0);
    xy -= y;
    y -= input[(i - period) + 1];
  }

  *output_idx_ref = output_idx;
}


----------------------------
***/


